File: internal/commit/combine.go
package commit

import (
    "errors"
    "fmt"
    "os"
    "strings"

    "github.com/diesi/aic/internal/cli"
    "github.com/diesi/aic/internal/config"
    "github.com/diesi/aic/internal/openai"
    "github.com/diesi/aic/internal/provider"
)

// GenerateCombinedSuggestions asks the AI to combine multiple commit messages
// into a fresh set of consolidated suggestions. It returns up to cfg.Suggestions
// items, formatted one message per choice with no numbering or bullets.
func GenerateCombinedSuggestions(cfg Config, apiKey string, selected []string) ([]string, error) {
	if len(selected) < 2 {
		return nil, errors.New("need at least two messages to combine")
	}
	if config.Bool(config.EnvAICMock) {
        fused := strings.Join(selected, "; ")
        out := []string{
            fused,
            "Refine combined suggestions for clarity",
            "Improve wording across merged changes",
            "Consolidate changes into a single subject",
            "Address edge cases from merged updates",
        }
        if cfg.Suggestions > 0 && cfg.Suggestions < len(out) {
            out = out[:cfg.Suggestions]
        }
        return out, nil
    }
    if apiKey == "" {
        switch cfg.Provider {
        case "claude":
            return nil, errors.New("missing CLAUDE_API_KEY")
        case "gemini":
            return nil, errors.New("missing GEMINI_API_KEY")
        case "custom":
            // Custom providers may not require a key (e.g., local LM Studio)
            // Proceed without error.
        default:
            return nil, errors.New("missing OPENAI_API_KEY")
        }
    }
	var p provider.Provider
    switch cfg.Provider {
    case "claude":
        p = provider.NewClaude(apiKey)
    case "gemini":
        p = provider.NewGemini(apiKey)
    case "custom":
        p = provider.NewCustom(apiKey)
    default:
        p = provider.NewOpenAI(apiKey)
    }
    systemMsg := "You synthesize multiple draft commit messages into improved, concise natural-language Git commit subjects. " +
        "Rules: output a single-line subject per choice, imperative mood, no trailing period; no type prefixes or scopes. " +
        "Prefer clarity even if longer than 92 characters. Return ONLY the subjects, with no numbering or bullets."
    if cfg.SystemAddition != "" {
        systemMsg += " Additional user instructions: " + cfg.SystemAddition
    }
    if config.Bool(config.EnvAICDebug) {
        fmt.Fprintln(os.Stderr, "[aic][debug] system prompt for combine:")
        fmt.Fprintln(os.Stderr, systemMsg)
    }
	userContent := "Combine and refine these commit messages into consolidated alternatives:\n\n" + strings.Join(selected, "\n")

    temp := float32(0.4)
    resp, err := p.Chat(openai.ChatCompletionRequest{
        Model:       cfg.Model,
        Messages:    []openai.Message{{Role: "system", Content: systemMsg}, {Role: "user", Content: userContent}},
        // Allow longer combined subjects when helpful
        MaxTokens:   512,
        N:           cfg.Suggestions,
        Temperature: &temp,
    })
	if err != nil {
		return nil, err
	}
	if len(resp.Choices) == 0 {
		return nil, errors.New("no choices returned")
	}
	suggestions := make([]string, 0, len(resp.Choices))
	for _, msg := range resp.Choices {
		msg = strings.TrimSpace(msg)
		if msg == "" {
			continue
		}
		lines := []string{msg}
		if strings.Contains(msg, "\n") {
			lines = []string{}
			for _, line := range strings.Split(msg, "\n") {
				line = strings.TrimSpace(line)
				if line == "" {
					continue
				}
				lines = append(lines, line)
			}
		}
		for _, ln := range lines {
			ln = cli.StripLeadingListMarker(ln)
			if ln == "" {
				continue
			}
			suggestions = append(suggestions, ln)
		}
	}
	if len(suggestions) == 0 {
		errMsg := "empty suggestions after combining"
		if config.Bool(config.EnvAICDebug) && resp != nil && resp.Raw != "" {
			errMsg = fmt.Sprintf("%s\n\nRaw Response:\n%s", errMsg, resp.Raw)
		}
		return nil, errors.New(errMsg)
	}
	if len(suggestions) > cfg.Suggestions {
		suggestions = suggestions[:cfg.Suggestions]
	}
	return suggestions, nil
}

File: src/commit/combine.ts
import { getApiKeyForProvider, newProviderClient, type ProviderName } from "../providers";
import { wrapCommitMessage } from "./wrap";
import { stripLeadingListMarker } from "./listmarker";
import { debugVerbose } from "../debug";
import { GENERATION_CONFIG } from "../constants";

export type CombineConfig = {
  provider: ProviderName;
  model: string;
  suggestions: number;
  systemAddition: string;
};

export async function generateCombinedSuggestions(cfg: CombineConfig, selected: string[]): Promise<string[]> {
  if (!selected || selected.length < 2) throw new Error("need at least two messages to combine");
  const apiKey = getApiKeyForProvider(cfg.provider);
  const client = newProviderClient(cfg.provider, apiKey);

  const system =
    "You synthesize multiple draft commit messages into improved, concise natural-language Git commit subjects. " +
    "Rules: write descriptive commit messages, imperative mood, no trailing period; no type prefixes or scopes. " +
    "Prioritize clarity and meaningful detail; describe what changed in specific terms. Do NOT include why changes were made or any rationale. " +
    "Preserve the strongest concrete noun phrases from the drafts (key modules, features, or components), but rewrite with fresh sentence structure and verb choices. " +
    "Do not simply paraphrase one draft; synthesize the shared signal across drafts into stronger subjects. " +
    "Identify the most impactful shared theme (largest scope, user-facing, architectural/security, or release-impacting) and prioritize that in the subject. " +
    "Produce distinct alternatives with different phrasing and emphasis (varied verbs, structures, focus). " +
    "If the message would exceed 80 characters, wrap ONLY between sentences; never break mid-sentence. If a sentence exceeds 80 characters, keep it on a single line. " +
    "Return ONLY the subjects, with no numbering or bullets." +
    (cfg.systemAddition ? " Additional user instructions: " + cfg.systemAddition : "");

  const user = "Combine and refine these commit messages into consolidated alternatives:\n\n" + selected.join("\n");
  debugVerbose("SYSTEM", `combine system prompt: ${system}`);
  debugVerbose("CONTENT", `combine user content length=${user.length}`);

  const maxTokens = 768;

  const messages = [
    { role: "system" as const, content: system },
    { role: "user" as const, content: user },
  ];

  const out: string[] = [];
  const seen = new Set<string>();
  const target = Math.max(1, cfg.suggestions);
  let attempts = 0;
  // First attempt: ask for as many as needed in one go (providers that support n)
  {
    const need = target;
    const temperature = jitterTemperature(
      0,
      GENERATION_CONFIG.COMBINE_TEMPERATURE_MIN,
      GENERATION_CONFIG.COMBINE_TEMPERATURE_MAX,
      GENERATION_CONFIG.COMBINE_TEMPERATURE
    );
    const resp = await client.chat({ model: cfg.model, messages, maxTokens, temperature, n: need });
    for (const raw of resp.choices) {
      for (const s of extractSuggestions(raw)) {
        const norm = s.toLowerCase();
        if (!seen.has(norm)) {
          seen.add(norm);
          out.push(s);
        }
        if (out.length >= target) break;
      }
      if (out.length >= target) break;
    }
  }
  // Refill individually if needed
  while (out.length < target && attempts < target * 3) {
    attempts++;
    const temperature = jitterTemperature(
      attempts,
      GENERATION_CONFIG.COMBINE_TEMPERATURE_MIN,
      GENERATION_CONFIG.COMBINE_TEMPERATURE_MAX,
      GENERATION_CONFIG.COMBINE_TEMPERATURE
    );
    const resp = await client.chat({ model: cfg.model, messages, maxTokens, temperature });
    for (const raw of resp.choices) {
      for (const s of extractSuggestions(raw)) {
        const norm = s.toLowerCase();
        if (!seen.has(norm)) {
          seen.add(norm);
          out.push(s);
        }
        if (out.length >= target) break;
      }
      if (out.length >= target) break;
    }
  }
  // Fallbacks: ensure at least one result
  if (!out.length) {
    // simple merged baseline
    const merged = selected.join("; ");
    out.push(postProcess(merged) || merged);
  }
  return out.slice(0, target);
}

function postProcess(text: string): string {
  const msg = (text || "").trim();
  if (!msg) return "";
  const lines = msg.includes("\n")
    ? msg
        .split("\n")
        .map((l) => l.trim())
        .filter(Boolean)
    : [msg];
  const out: string[] = [];
  for (let ln of lines) {
    ln = stripLeadingListMarker(ln);
    if (!ln) continue;
    out.push(wrapCommitMessage(ln));
  }
  return out[0] || "";
}

function extractSuggestions(text: string): string[] {
  const res: string[] = [];
  const lines = (text || "")
    .split("\n")
    .map((l) => l.trim())
    .filter(Boolean);
  for (let ln of lines) {
    ln = stripLeadingListMarker(ln);
    if (!ln) continue;
    const wrapped = wrapCommitMessage(ln);
    if (wrapped) res.push(wrapped);
  }
  // If the model returned a single paragraph, take it as one suggestion
  if (!res.length) {
    const s = postProcess(text);
    if (s) res.push(s);
  }
  return res;
}

function jitterTemperature(seed: number, min: number, max: number, fallback: number): number {
  if (!Number.isFinite(min) || !Number.isFinite(max) || max <= min) return fallback;
  const value = pseudoRandom01(seed);
  const temp = min + value * (max - min);
  return Number(temp.toFixed(2));
}

function pseudoRandom01(seed: number): number {
  const x = Math.sin((seed + 1) * 12.9898) * 43758.5453;
  return x - Math.floor(x);
}
